File: server/datasets/python_source-code/Multiple-D1/organic_12.py
"""
Quick Sort - Organic Submission #12
Tail-call optimized Quick Sort (recurse on smaller half, loop on larger).
"""

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def quick_sort_tail(arr, low, high):
    while low < high:
        pi = partition(arr, low, high)
        if pi - low < high - pi:
            quick_sort_tail(arr, low, pi - 1)
            low = pi + 1
        else:
            quick_sort_tail(arr, pi + 1, high)
            high = pi - 1

if __name__ == "__main__":
    nums = [10, 7, 8, 9, 1, 5]
    quick_sort_tail(nums, 0, len(nums) - 1)
    print(nums)

File: server/datasets/java_source-code/Multiple-D1/organic_12.java
/*
 * Quick Sort - Organic Submission #12
 * Tail-call optimized Quick Sort (recurse on smaller half, loop on larger).
 */

class Solution {
    public static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i]; arr[i] = arr[j]; arr[j] = temp;
            }
        }
        int temp = arr[i + 1]; arr[i + 1] = arr[high]; arr[high] = temp;
        return i + 1;
    }

    public static void quickSortTail(int[] arr, int low, int high) {
        while (low < high) {
            int pi = partition(arr, low, high);
            if (pi - low < high - pi) {
                quickSortTail(arr, low, pi - 1);
                low = pi + 1;
            } else {
                quickSortTail(arr, pi + 1, high);
                high = pi - 1;
            }
        }
    }

    public static void main(String[] args) {
        int[] nums = {10, 7, 8, 9, 1, 5};
        quickSortTail(nums, 0, nums.length - 1);
        System.out.println(java.util.Arrays.toString(nums));
    }
}
